### ISSUE TITLE
Medium-term issue: using a proc-macro for better shaders

### ISSUE THREAD

**Issue by tomaka** [vulkano-rs/vulkano#549]
Right now you can use macros 1.1 to put shaders inside of your source code and compile them at compile-time. When doing so, vulkano will generate the Rust equivalents of the structs used in the GLSL code. Whenever you declare a struct in GLSL, vulkano will generate its equivalent in Rust.

However these generated structs have a big problem: if multiple shaders declare the same struct (which happens often), they will each generate their own version of that struct.
For example if shader 1 declares a struct named `Foo`, and shader 2 declares the same `Foo`, then two different `Foo` structs will be generated. Even though their names and declarations are strictly identical, the Rust compiler will still think that they are distinct.
This means that vulkano can't use these generated structs for type safety. If vulkano required that the user passes shader1's `Foo` when invoking shader1 and shader2's `Foo` when invoking shader2, then there would be no way to invoke shaders 1 and 2 at the same time (which, again, happens often, eg. if shader 1 is a vertex shader and shader 2 is a fragment shader).

The solutions to this problem proposed here is to ask users to write themselves the structs in Rust, and either import these Rust structs into the GLSL code, or check whether the structs declared in the GLSL code match the Rust structs.

Eventually I think it should be possible for a vulkano-using app to write for example something like this:

## Solution 1

```rust
#[repr(C)]
#[gpu]
struct PushConstants {
    a: u32,
}

shader!{ MyShader, "
#version 450
#vulkano_import_struct PushConstants;

layout(push_constant) uniform PushConstants push_constants;

void main() {
    // ...
}
"};
```

By running a procedural macro on this code, we can analyze the layout of the Rust `PushConstants` struct*. Then the compiler would generate the GLSL equivalent of this struct declaration in the GLSL code before passing it to the compiler. In practice I think the best thing to do is to simply return an error if the Rust struct doesn't match either the std140 or std430 layout.

*: I don't know if it's actually possible to determine or to ask the compiler the layout that a struct is going to have, but we could easily force a `#[repr(C)]` attribute on the struct.

Alternatively we could ask the user to write this:

## Solution 2

```rust
#[repr(C)]
#[gpu]
struct PushConstants {
    a: u32,
}

shader!{ MyShader, "
#version 450

layout(push_constant) uniform PushConstants {
    uint a;
} push_constants;

void main() {
    // ...
}
"};
```

The user would have to declare the struct both in Rust and GLSL.

On one hand this is less magic than solution 1 because we don't introduce any preprocessing directive, but on the other hand it is more magic because we assume that the `PushContants` struct is in scope. I don't really see how to handle a situation where the user would declare the Rust struct in a module and write the GLSL code in a different module.

Anyway, both these solutions would be a good first step towards #353.

This issue talks about struct layouts, but exactly the same could be done for descriptor sets.

**Comment by YeyaSwizaw** [comment]
I have been working on a simple attempt at fixing at least the code duplication problem, by adding a #[derive(VulkanoStruct)] to the proc macro crate. This does no code generation (at the moment), but simply parses the struct and stores information in a lazy_static hashmap. The shader derive macro then simply scans the source for #[vulkano_struct(<name>)] and replaces it with a struct definition from the hashmap.

As an example:
```rust
#[repr(C)]
#[derive(VulkanoStruct)]
struct Uniforms {
    pos: [u32; 2]
}

mod shader {
    #[derive(VulkanoShader)]
    #[ty = "vertex"]
    #[src = "
#version 450

layout(set = 0, binding = 0) uniform #[vulkano_struct(Uniforms)] uniforms;

layout(location = 0) in vec2 point;

void main() {
    gl_Position = vec4(uniforms.pos * point, 0, 1);
}
    "]
    struct Dummy;
}
```

The `#[vulkano_struct(Uniforms)]` will be replaced by `Uniforms { uvec2 pos; }` before the source is sent for compilation.

One idea for then expanding this for type-safety could be to generate a marker trait of some sort with each struct, and then require that trait when using the shader.

**Comment by tomaka** [comment]
That's nice!

> This does no code generation (at the moment), but simply parses the struct and stores information in a lazy_static hashmap.

I don't think this is a robust solution though. If the compiler invokes the derive of the shader before the derive of the struct, then there is a problem.

**Comment by YeyaSwizaw** [comment]
Unfortunately, with proc macros as they are, I can think of no other way of reasonably getting information about the structs.

Even going for solution 2, which could be implemented easily by passing a path to the macro (`#[struct = "Uniforms = ::Uniforms"]` or something), but we'd have no way of verifying that the path exists or that the struct is in the correct format.

**Comment by Rua** [comment]
I believe this was implemented by #1694.
